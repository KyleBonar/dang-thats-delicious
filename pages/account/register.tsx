import * as React from "react";
import RegisterUser from "../../src/components/RegisterUser/RegisterUser";
import { Article } from "../../src/components/Article/Article";
import HeaderSimple from "../../src/components/HeaderSimple/HeaderSimple";

export interface RegisterProps {}

const Register: React.SFC<RegisterProps> = () => {
  return (
    <>
      <HeaderSimple />
      <Article size="sm">
        <RegisterUser />
      </Article>
    </>
  );
};

export default Register;
