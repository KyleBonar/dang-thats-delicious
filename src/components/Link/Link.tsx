import * as React from "react";
import Link from "next/link";
import styled from "../../theme/styled-components";

interface LinkProps {
  children:
    | string
    | Element
    | JSX.Element
    | Array<string | Element | JSX.Element>;
  to?: string;
  href?: string;
  className?: string;
  target?: "_blank" | "_self";
}

const LinkComponent: React.FC<LinkProps> = props => {
  return (
    <>
      {props.target === "_blank" ? (
        <a
          href={props.to || props.href || "#"}
          target="_blank"
          className={props.className}
        >
          {props.children}
        </a>
      ) : (
        <Link href={props.to || props.href || "#"}>
          <a className={props.className} target="_self">
            {props.children}
          </a>
        </Link>
      )}
    </>
  );
};

const StyledLink = styled(LinkComponent)`
  text-decoration: none;
  font-family: FuturaMedium;
  color: ${props => props.theme.primaryThemeColor};

  &:hover,
  &:focus {
    color: ${props => props.theme.secondaryThemeColor};
  }
`;

export { StyledLink as Link };
