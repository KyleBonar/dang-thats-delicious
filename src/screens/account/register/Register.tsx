import * as React from "react";
import { connect } from "react-redux";

import { register } from "../../../redux/users/actions";
import { IRegisterUser } from "../../../redux/users/types";
import LogoSFYT from "../../../images/icons/LogoSFYT";
import PageTitle from "../../../components/PageTitle/PageTitle";
import { TextInput } from "../../../components/TextInput/TextInput";
import { Link } from "../../../components/Link/Link";
import { AppState } from "../../../redux/configureStore";
import {
  StyledDiv,
  StyledLogoContainer,
  StyledArticle,
  StyledFormContainer,
  StyledText,
  StyledButton
} from "./RegisterStyle";
import {
  FlashMessage,
  FlashMessageProps
} from "../../../components/FlashMessage/FlashMessage";
import Auth from "../../../utils/Auth/Auth";

export interface RegisterProps {
  register: any;
  history: { push: (location: string) => null };
}

export interface RegisterState {
  email: string;
  confirmEmail: string;
  password: string;
  confirmPassword: string;
  displayName: string;
  flashMessage: FlashMessageProps;
}

class Register extends React.Component<RegisterProps, RegisterState> {
  constructor(props: RegisterProps) {
    super(props);

    // Init state
    this.state = {
      email: "",
      confirmEmail: "",
      password: "",
      confirmPassword: "",
      displayName: "",
      flashMessage: {
        isVisible: false
      }
    };
  }

  public render() {
    return (
      <StyledDiv>
        <StyledLogoContainer>
          <Link to="/">
            <LogoSFYT />
          </Link>
        </StyledLogoContainer>
        <hr />
        <StyledArticle>
          <PageTitle>Register</PageTitle>
          <StyledFormContainer>
            {this.state.flashMessage.isVisible && (
              <FlashMessage type={this.state.flashMessage.type} isVisible>
                {this.state.flashMessage.text}
              </FlashMessage>
            )}
            <form onSubmit={this.onSubmit} style={{ width: "100%" }}>
              <TextInput
                type="email"
                onChange={this.onTextChange}
                showLabel={true}
                label={"Email"}
                name={"email"}
                value={this.state.email}
                required={true}
              />
              <TextInput
                type="email"
                onChange={this.onTextChange}
                showLabel={true}
                label={"Confirm Email"}
                name={"confirmEmail"}
                value={this.state.confirmEmail}
                required={true}
                requirementText={"Must match above."}
              />
              <TextInput
                type="password"
                onChange={this.onTextChange}
                showLabel={true}
                label={"Password"}
                name={"password"}
                value={this.state.password}
                required={true}
                requirementText={"Must be at least 9 characters long."}
              />
              <TextInput
                type="password"
                onChange={this.onTextChange}
                showLabel={true}
                label={"Confirm Password"}
                name={"confirmPassword"}
                value={this.state.confirmPassword}
                required={true}
                requirementText={"Must match above."}
              />
              <TextInput
                type="text"
                onChange={this.onTextChange}
                showLabel={true}
                label={"Display Name"}
                name={"displayName"}
                value={this.state.displayName}
                required={true}
                requirementText={"Must be at least 6 characters long."}
              />
              <StyledText>
                By clicking 'Register', you agree to Sauce For Your Thoughts{" "}
                <Link to="#">Terms and Conditions</Link>
              </StyledText>
              <StyledButton type="submit">Register</StyledButton>
            </form>
          </StyledFormContainer>
        </StyledArticle>
      </StyledDiv>
    );
  }

  private onTextChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    if (!event || !event.target) {
      return;
    }
    // Grab the name and value
    const { name, value }: { name: string; value: string } = event.target;

    // Update local state
    this.setState({
      ...this.state,
      [name]: value
    });
  };

  private onSubmit = async (event: React.FormEvent): Promise<any> => {
    event.preventDefault();

    if (this.state.email !== this.state.confirmEmail) {
      this.setState({
        flashMessage: {
          isVisible: true,
          text: "Your emails do not match. Please fix this before continuing.",
          type: "alert"
        }
      });

      window.scrollTo(0, 0); // Move screen to top

      return;
    }

    if (this.state.password !== this.state.confirmPassword) {
      this.setState({
        flashMessage: {
          isVisible: true,
          text:
            "Your passwords do not match. Please fix this before continuing.",
          type: "alert"
        }
      });

      window.scrollTo(0, 0); // Move screen to top

      return;
    }

    if (this.state.password.length < 8) {
      this.setState({
        flashMessage: {
          isVisible: true,
          text:
            "Your password is too weak! Please make your password over 8 characters long.",
          type: "alert"
        }
      });

      window.scrollTo(0, 0); // Move screen to top

      return;
    }

    if (this.state.displayName.length < 6) {
      this.setState({
        flashMessage: {
          isVisible: true,
          text:
            "Your display name is not long enough. It must be at least 6 characters long.",
          type: "alert"
        }
      });

      window.scrollTo(0, 0); // Move screen to top

      return;
    }

    // Grab values from state
    const {
      email,
      confirmEmail,
      password,
      confirmPassword,
      displayName
    } = this.state;

    // Create credentials obj
    const credentials: IRegisterUser = {
      user: { email, confirmEmail, password, confirmPassword, displayName }
    };
    try {
      // dispatch action which calls API to register user
      await this.props.register({ credentials });

      // Redirect user to sauces page -- Maybe take them to user home page instead?
      this.props.history.push("/sauces");
    } catch (err) {
      // Create warning flash
      this.setState({
        flashMessage: {
          isVisible: true,
          text: err.response.data.msg,
          type: "warning"
        }
      });

      window.scrollTo(0, 0); // Move screen to top
    }
  };
}

const mapState2Props = (state: AppState) => {
  return {};
};

const mapDispatch2Props = {
  register
};

export default connect(
  mapState2Props,
  mapDispatch2Props
)(Register);
