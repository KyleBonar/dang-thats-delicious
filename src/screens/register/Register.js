import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";

import RegisterForm from "./components/registerForm/RegisterForm";

import { flashError, flashClose } from "../../redux/actions/flash";
import { register } from "../../redux/actions/auth";

class Register extends Component {
  static propType = {
    history: PropTypes.shape({ push: PropTypes.func.isRequired }).isRequired,
    register: PropTypes.func.isRequired,
    flashError: PropTypes.func.isRequired,
    flashClose: PropTypes.func.isRequired
  };
  render() {
    return (
      <div className="inner">
        <RegisterForm onSubmit={this.handleSubmit} />
      </div>
    );
  }

  handleSubmit = e => {
    const data = {
      user: {
        email: e.email,
        name: e.name,
        password: e.password,
        confirmPassword: e.confirmPassword
      }
    };
    this.props.flashClose();
    this.props
      .register(data)
      .then(() => this.props.history.push("/sauces"))
      .catch(err => {
        console.log(err);
        console.log(err.response);
        this.props.flashError({ text: err.response.data.msg });
      });
  };
}

const mapDispatchToProps = { register, flashError, flashClose };

export default connect(null, mapDispatchToProps)(Register);
