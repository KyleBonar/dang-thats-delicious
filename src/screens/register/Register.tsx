import * as React from "react";
import { connect } from "react-redux";
import LogoSFYT from "../../images/icons/LogoSFYT";
import styled from "styled-components";
import Article from "../../components/Article/Article";
import PageTitle from "../../components/PageTitle/PageTitle";
import TextInput from "../../components/TextInput/TextInput";

const StyledLogoContainer = styled.div`
  max-width: 150px;
  margin: 0 auto;
  padding: 1em;
`;

const StyledArticle = styled(Article)`
  max-width: 600px;
`;
const StyledFormContainer = styled.div`
  background-color: ${props => props.theme.formContainerBackgroundColor};
  padding: 1.5rem;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
`;

export interface RegisterProps {}

export interface RegisterState {}

class Register extends React.Component<RegisterProps, any> {
  public render() {
    return (
      <div>
        <StyledLogoContainer>
          <LogoSFYT />
        </StyledLogoContainer>
        <hr />
        <StyledArticle>
          <PageTitle>Register</PageTitle>
          <StyledFormContainer>
            <TextInput
              type="text"
              onChange={() => {}}
              showLabel={true}
              label={"Email"}
              required={true}
            />
            <TextInput
              type="text"
              onChange={() => {}}
              showLabel={true}
              label={"Confirm Email"}
              required={true}
            />
            <TextInput
              type="text"
              onChange={() => {}}
              showLabel={true}
              label={"Password"}
              required={true}
            />
            <TextInput
              type="text"
              onChange={() => {}}
              showLabel={true}
              label={"Confirm Password"}
              required={true}
            />
            <TextInput
              type="text"
              onChange={() => {}}
              showLabel={true}
              label={"Display Name"}
              required={true}
            />
          </StyledFormContainer>
        </StyledArticle>
      </div>
    );
  }
}

const mapState2Props = state => {
  return {};
};

export default connect(mapState2Props)(Register);
