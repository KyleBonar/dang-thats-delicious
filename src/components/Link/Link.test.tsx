import "jsdom-global/register";
import * as React from "react";
import * as enzyme from "enzyme";

import { Link, LinkProps } from "./Link";
import {
  casual,
  fakeJSXElement,
  ITERATION_SIZE
} from "../../utils/testUtils/testUtils";

const mockLink = (): LinkProps => ({
  children: casual.random_element([
    casual.string,
    fakeJSXElement(),
    [casual.string, fakeJSXElement()]
  ]),
  href: casual.url,
  className: casual.boolean ? casual.string : undefined,
  target: casual.boolean
    ? casual.random_element(["_blank", "_self"])
    : undefined
});

const mockLinks = new Array(ITERATION_SIZE).fill(null).map(mockLink);

describe("<Link>", () => {
  it("renders", () => {
    const wrapper = enzyme.shallow(<Link href="#">""</Link>);
    expect(wrapper).toBeTruthy();
  });

  it("matches snapshot", () => {
    mockLinks.forEach((mockLink: LinkProps) => {
      const wrapper = enzyme.shallow(
        <Link {...mockLink}>{mockLink.children}</Link>
      );

      expect(wrapper).toMatchSnapshot();
    });
  });

  it("renders correct children", () => {
    mockLinks.forEach((mockLink: LinkProps) => {
      const wrapper = enzyme.mount(
        <Link {...mockLink}>{mockLink.children}</Link>
      );

      expect(wrapper.find("a").props().children).toEqual(mockLink.children);
    });
  });

  it("renders correct link", () => {
    mockLinks.forEach((mockLink: LinkProps) => {
      const wrapper = enzyme.mount(
        <Link {...mockLink}>{mockLink.children}</Link>
      );

      expect(wrapper.find("a").props().href).toEqual(mockLink.href);
    });
  });

  it("accepts additional classname", () => {
    mockLinks.forEach((mockLink: LinkProps) => {
      if (!mockLink.className) return;

      const wrapper = enzyme.mount(
        <Link {...mockLink}>{mockLink.children}</Link>
      );

      expect(wrapper.find("a").props().className).toContain(mockLink.className);
    });
  });

  it("render outward link if asked otherwise default", () => {
    mockLinks.forEach((mockLink: LinkProps) => {
      const wrapper = enzyme.mount(
        <Link {...mockLink}>{mockLink.children}</Link>
      );

      expect(wrapper.find("a").props().target).toContain(
        mockLink.target || "_self"
      );
    });
  });
});
