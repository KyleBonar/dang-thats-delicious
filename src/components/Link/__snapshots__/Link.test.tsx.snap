// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link> matches snapshot 1`] = `
<LinkComponent
  className="sc-AxjAm kGmApj"
  href="http://Ashton.org/"
>
  <div
    key="2d9e9bcc-f46a-4516-8896-4ffea6ae8c41"
  >
    autem et modi tenetur amet modi reprehenderit
  </div>
</LinkComponent>
`;

exports[`<Link> matches snapshot 2`] = `
<LinkComponent
  className="sc-AxjAm kGmApj"
  href="http://www.Bradley.io/"
  target="_blank"
>
  libero odit aut labore qui aut inventore
  <span
    key="bbc36e8a-3928-4077-baa2-9c927928a031"
  >
    molestiae cupiditate autem consequatur consequatur accusantium molestiae
  </span>
</LinkComponent>
`;

exports[`<Link> matches snapshot 3`] = `
<LinkComponent
  className="sc-AxjAm kGmApj autem eligendi nihil quis eveniet sit voluptas"
  href="http://www.Botsford.biz/"
  target="_self"
>
  <span
    key="d82ec38a-3d7c-4be5-9157-537ce2045691"
  >
    accusantium vel enim sint voluptatem porro ut
  </span>
</LinkComponent>
`;

exports[`<Link> matches snapshot 4`] = `
<LinkComponent
  className="sc-AxjAm kGmApj"
  href="http://www.Jaron.name/"
  target="_blank"
>
  <span
    key="c664e88b-186b-4bcd-83b1-37fd3a9e071a"
  >
    reiciendis perspiciatis ut illo sit rerum accusamus
  </span>
</LinkComponent>
`;

exports[`<Link> matches snapshot 5`] = `
<LinkComponent
  className="sc-AxjAm kGmApj et debitis dolorem possimus accusantium enim est"
  href="http://www.Morissette.name/"
  target="_blank"
>
  <div
    key="9bfe3d66-49bc-4a9a-ac27-7289385c16b5"
  >
    delectus aliquid et dicta ipsam qui provident
  </div>
</LinkComponent>
`;

exports[`<Link> matches snapshot 6`] = `
<LinkComponent
  className="sc-AxjAm kGmApj possimus voluptas eaque expedita iste velit et"
  href="http://www.Upton.tv/"
>
  <div
    key="9babf7d5-46ce-4172-9608-15c773cff13b"
  >
    quia blanditiis qui sed a dolores voluptatem
  </div>
</LinkComponent>
`;

exports[`<Link> matches snapshot 7`] = `
<LinkComponent
  className="sc-AxjAm kGmApj rerum aliquid reiciendis dolorem porro voluptates ut"
  href="http://www.Maude.name/"
  target="_blank"
>
  laborum officia iure eius veritatis voluptates suscipit
</LinkComponent>
`;

exports[`<Link> matches snapshot 8`] = `
<LinkComponent
  className="sc-AxjAm kGmApj"
  href="http://McDermott.biz/"
>
  repellendus qui est quis est veniam eum
</LinkComponent>
`;
